examples/multisig_plain/lib.rs

```diff
@@ -0,0 +1,837 @@
+            self.env().emit_event(OwnerAddition { owner: new_owner });
+        }
+
+        /// Change the requirement to a new value.
+        /// Only callable by the wallet itself.
+        #[ink(message)]
+        fn change_requirement(&mut self, new_requirement: u32) {
+            self.ensure_from_wallet();
+            self.ensure_requirement_is_valid(self.owners.len(), new_requirement);
+            self.requirement.set(new_requirement);
+            self.env().emit_event(RequirementChange { new_requirement });
+        }
+
+        /// Add a new transaction candiate to the contract.
+        /// This also confirms the transaction for the caller.
+        /// This can be called by any owner.
+        #[ink(message)]
+        fn submit_transaction(&mut self, transaction: Transaction) {
+            self.ensure_caller_is_owner();
+            let trans_id = self.transactions.put(transaction);
+            self.confirmation_count.insert(trans_id, 0);
+            self.env().emit_event(Submission {
+                transaction: trans_id,
+            });
+            self.confirm_by_caller(self.env().caller(), trans_id);
+        }
+
+        /// Remove a transaction from the contract.
+        /// Only callable by the wallet itself.
+        /// Panics if `trans_id` is no valid transaction id.
+        #[ink(message)]
+        fn cancel_transaction(&mut self, trans_id: TransactionId) {
+            self.ensure_from_wallet();
+            if self.take_transaction(trans_id).is_some() {
+                self.env().emit_event(Cancelation {
+                    transaction: trans_id,
+                });
+            }
+        }
+
+        /// Confirm a transaction for the sender that was submitted by any owner.
+        /// This can be called by any owner.
+        /// Panics if `trans_id` is no valid transaction id.
+        #[ink(message)]
+        fn confirm_transaction(&mut self, trans_id: TransactionId) {
+            self.ensure_caller_is_owner();
+            self.ensure_transaction_exists(trans_id);
+            self.confirm_by_caller(self.env().caller(), trans_id);
+        }
+
+        /// Revoke the senders confirmation.
+        /// This can be called by any owner.
+        /// Panics if `trans_id` is no valid transaction id.
+        #[ink(message)]
+        fn revoke_confirmation(&mut self, trans_id: TransactionId) {
+            self.ensure_caller_is_owner();
+            let caller = self.env().caller();
+            if self.confirmations.remove(&(trans_id, caller)).is_some() {
+                mutate_map(&mut self.confirmation_count, &trans_id, |count| *count -= 1);
+                self.env().emit_event(Revokation {
+                    transaction: trans_id,
+                    from: caller,
+                });
+            }
+        }
+
+        /// Execute a confirmed execution.
+        /// Its return type indicates whether the called transaction was succesful.
+        /// This can be called by anyone.
```

Comment:
Does the original multi-sig return the result of the called function? Because this implementation doesn't but I think we could return some bytes back to the caller. This, however, should be another call since that's more work done. What do you think? Or should we keep it simple?

Reply:
Ethereum contracts cannot return anything AFAIK. You completely rely on events there. Therefore you would detect the failure by listening for the callees events.

I think you are right. We should make use of return values as its possible in substrate.

If you put this in a separate call you must store the result in contract storage. I do not know if this would be so good.

Reply:
I split `execute_transaction` into `eval_transaction` and `invoke_transaction`. The former one retrieves the output of the call as bytes.